### PR TITLE
Add connections pane support for structured object types

### PR DESCRIPTION
**Pull request by jmcphers**
Prior to this change, the connections pane supported only tables. After the change, it supports a hierarchy of objects, as supplied by the connection itself. The goal is to make it possible to navigate large databases by allowing users to drill down to specific objects as they do in traditional DBMS software.

image

There are three main changes which support this feature:

1. A connection now emits its object hierarchy when initialized, and the hierarchy is stored along with the connection. See the new `listObjectTypes` method on the connection observer, and the new `object_types` member of the persisted connection.

2. When specifying an object to interact with (e.g. preview), a fully qualified specifier is now necessary (not just a string with the object's name). This specifier is an array of names and types giving the parents of the object; for instance, a table `humanresources.employee` might be specified as `[{name = "humanresources", type = "schema"}, {name = "employee", type = "table" }]`. These become the arguments to the functions supplied by the connection, and are opaque to RStudio.

3. The object browser is now capable of showing several levels of nested objects; each object has its own child object provider which knows how to list its children (using the stack of specifiers mentioned above). 

There are pull requests into both the ODBC and SparklyR packages which use this new contract:

https://github.com/rstats-db/odbc/pull/42

https://github.com/rstudio/sparklyr/pull/580

### CODE REVIEW

**Comment by jjallaire** [comment]
@jeroen This is a new API/protocol for packages to register themselves as a "Connection Provider"  with the IDE. This primes the new **Connections** pane to:

1. Provide a UX to assist in establishing connections

2. Recall previous code used to establish connections

3. Provide a interface for browsing the contents of the connection and viewing it's objects/tables.

I'm hoping that **mongolite** might be able to take advantage of this. Take a look and let us know what you think.

File: src/cpp/session/modules/connections/SessionConnections.cpp [hunk 1]
```diff
@@ -74,7 +76,7 @@ SEXP rs_connectionOpened(SEXP connectionSEXP)
    if (!r::sexp::isNull(actionList))
    {
       std::vector<std::string> actionNames;
-      error = r::sexp::getNames(actionList, &actionNames);
+      r::sexp::getNames(actionList, &actionNames);
```
Comment (kevinushey) [inline src/cpp/session/modules/connections/SessionConnections.cpp]: Is it intentional that you're no longer capturing a potential error here?

Comment (jmcphers) [inline src/cpp/session/modules/connections/SessionConnections.cpp]: Yes, that error turned out to be too noisy -- if we can't extract the list of names it's fine as it just means we won't have any actions. Arguably some feedback for the package developer might be in order but that probably doesn't belong in our server logs. :-) (Elsewhere, errors which are severe enough to render the connection unusable should indeed be logged).

Comment (kevinushey) [inline src/cpp/session/modules/connections/SessionConnections.cpp]: Gotcha -- I think the following `if (error) LOG_ERROR(error);` bit could be stripped out then; does that sound right?

Comment (jmcphers) [inline src/cpp/session/modules/connections/SessionConnections.cpp]: It does, thanks for catching that! https://github.com/rstudio/rstudio/commit/4246a44ca2c94281cebfb207fcb9f4b5b718a1f1.

**Comment by jeroen** [comment]
Is this API described somewhere, or should I just look at examples from `odbc` / `sparklyr` ?

**Comment by jmcphers** [comment]
No, there isn't documentation for the API yet. This PR into odbc has the most complete example; I've just pushed a commit that describes some of the return values in greater detail.

https://github.com/rstats-db/odbc/pull/42

**Pull request merged**